src/nupic/engine/Input.cpp

```diff
@@ -537,6 +534,7 @@ const std::vector<std::vector<size_t>> &Input::getSplitterMap() const {
 
 template <typename T>
 void Input::getInputForNode(size_t nodeIndex, std::vector<T> &input) const {
+  // NOTE: nodeIndex is always 1 because we no longer support nodes.
```

Comment:
> // NOTE: nodeIndex is always 1 because we no longer support nodes.

This was a comment explaining what I found out but I am not so sure now.  Here is what I am finding:

- If you look at all of the regions implemented in python you will see that singleNodeOnly = true in the Spec structure for all of the region implementations.  The only exceptions that I found were in C++, namely TestNode.cpp, VectorFileEffector, and VectorFileSensor.   
- What that flag means is to disable the ability to set dimensions on regions or links.  The 'node' in the comment refers to the first dimension or row of at least a two dimensional array for an input buffer.  
- The getInputForNode( ) function takes the input buffer and re-maps the bits using the splitter map, calling once for each row in the input buffer.  Without dimensions there is only one row and the splitter maps are not very useful...not sure they can even be created.
- The splitter maps are the product of the LinkPolicy and its subclasses TestFanIn2LinkPolicy and UniformLinkPolicy.  In my opinion a link policy is too complicated to setup and there is almost no documentation on how to do it.
- So at the moment, there can be multple 'nodes' if dimensions are defined and splitter maps do work when calling the getInputForNode( ) function.

What I wanted to do was remove dimensions and all LinkPolicy logic.  This would leave the feature of output buffers to be connected to input buffers with no remapping of bits.  Fan-In would just append output buffers side-by-side in the input buffer.  Fan-out would not be affected.  Propogation delay would not be affected.  I actually coded this all up about a year ago.

However, with the SDR it seems that dimensions are back.  So I am hesitant to rip out dimensions from the Links, just in case we need dimensions for the SDRs.  So, I made sure that LinkPolicy and splitter maps also still work with it.

The question is...(and I have asked this before)... should I rip dimensions, LinkPolicy, and splitter maps out of Links and make everything much simpler?  When I asked the question on the forum I did not get a definite answer because I don't think people even know what that means.

So, until told otherwise I will leave them all working... and I will remove that Note.

Reply:
This is a useful memo, thank you! 

> What I wanted to do was remove dimensions and all LinkPolicy logic. This would leave the feature of output buffers to be connected to input buffers with no remapping of bits. Fan-In would just append output buffers side-by-side in the input buffer. Fan-out would not be affected. Propogation delay would not be affected. I actually coded this all up about a year ago. 

I would like to see such simplification in the API. @ctrl-z-9000-times also complained it's overcomplicated. 

> However, with the SDR it seems that dimensions are back. So I am hesitant to rip out dimensions from the Links, just in case we need dimensions for the SDRs. So, I made sure that LinkPolicy and splitter maps also still work with it. 

COuld be always use 1D view of SDRs for that? Or, I'll reiterate on the issue of Sparse vs SDR types, Sparse is not affected by this, right? So do it (Sparse-SDR one becomes obsoleted) the other way around and keep Sparse? (But that would lose some nice benefits of using SDR, as no-copy?)

Reply:
> I'll reiterate on the issue of Sparse vs SDR types, Sparse is not affected by this, right?

Not sure what you are asking.  The splitter maps do not make since unless the data is the Dense format of SDR data (with the nupic definition of SDR).  Other types of data such as reset, ids, or numeric data that is not zero/non-zero would not use the splitter maps.  Sparse data format (just the indexes) also would not make since with a splitter map unless it was converted to dense format.  However by using the SDR class it will do the right thing using the Dense data in the class.